ansible_base/models/resource.py

```diff
@@ -0,0 +1,122 @@
+import uuid
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.functions import Concat
+from django.db.utils import ProgrammingError
+
+from .service_id import service_id
+
+_resource_registry = None
+
+
+# Circular imports make this really hard to initialize correctly.
+# I don't know why, but I can't put this function anywhere else without inducing
+# circular imports.
+def get_registry():
+    global _resource_registry
+    if _resource_registry is None:
+        from django.conf import settings
+
+        if hasattr(settings, "RESOURCE_REGISTRY_CONFIG"):
+            from django.utils.module_loading import import_string
+
+            _resource_registry = import_string(settings.RESOURCE_REGISTRY_CONFIG)
+        else:
+            _resource_registry = False
+    return _resource_registry
+
+
+class ResourceManager(models.Manager):
+    """
+    Ansible IDs are made up of two parts: the service short ID and a resource
+    UUIDv4. For efficiency reasons, these are stored as two separate fields:
+        - _service_id: Service that the resource belongs to.
+        - resource_id: Resource UUID.
+
+    Since Ansible ID is the value that will be referenced in the API, this
+    manager automatically annotates ansible_id as <service_short_id>:<resource_id>
+    for easier lookups.
+    """
+
+    def get_queryset(self):
+        # this needs to wait for migrations to run before it becomes available.
+        try:
+            s_id = service_id().split('-')[0]
+        except ProgrammingError:
+            s_id = "00000000"
+
+        return (
+            super()
+            .get_queryset()
+            .annotate(
+                _computed_service_id=models.Case(
+                    models.When(_service_id__isnull=False, then=models.F('_service_id')), default=models.Value(s_id), output_field=models.CharField()
+                ),
+                _ansible_id=Concat("_computed_service_id", models.Value(":"), models.F("resource_id"), output_field=models.CharField()),
```

Comment:
I'd be really nice if we could make this a constant from somewhere. I can't image we'd want to change it but if we had to searching for `:` in the source is gonna be cringe.